File: app/src/main/java/tech/pauly/findapet/shared/BaseViewModel.java
package tech.pauly.findapet.shared;

import android.arch.lifecycle.Lifecycle;
import android.arch.lifecycle.LifecycleObserver;
import android.arch.lifecycle.OnLifecycleEvent;

import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;

public class BaseViewModel implements LifecycleObserver {

    private CompositeDisposable lifecycleSubscriptions = new CompositeDisposable();

    protected void subscribeOnLifecycle(Disposable subscription) {
        lifecycleSubscriptions.add(subscription);
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_PAUSE)
    private void disposeSubscriptions() {
        lifecycleSubscriptions.clear();
    }
}

File: app-old/src/main/java/tech/pauly/old/shared/BaseViewModel.kt
package tech.pauly.old.shared

import androidx.databinding.BaseObservable
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.OnLifecycleEvent
import io.reactivex.Completable
import io.reactivex.Observable
import io.reactivex.Single
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.disposables.Disposable

open class BaseViewModel : BaseObservable(), BaseLifecycleViewModel  {

    private val lifecycleSubscriptions = CompositeDisposable()

    @OnLifecycleEvent(Lifecycle.Event.ON_STOP)
    private fun disposeSubscriptions() {
        lifecycleSubscriptions.clear()
    }

    //region Extensions
    protected fun Disposable.onLifecycle() {
        lifecycleSubscriptions.add(this)
    }

    protected inline fun <T> Observable<T>.quickSubscribe(crossinline onNext: (T) -> Unit) {
        this.subscribe({ onNext(it) }, Throwable::printStackTrace)
                .onLifecycle()
    }

    protected inline fun <T> Single<T>.quickSubscribe(crossinline onNext: (T) -> Unit) {
        this.subscribe({ onNext(it) }, Throwable::printStackTrace)
                .onLifecycle()
    }

    protected inline fun Completable.quickSubscribe(crossinline onComplete: () -> Unit) {
        this.subscribe({ onComplete() }, Throwable::printStackTrace)
                .onLifecycle()
    }
    //endregion
}
